source/types.ts

```diff
@@ -0,0 +1,292 @@
+ * @param optionsUsed {Options} - The options used to set up the middleware
+ */
+export type RateLimitExceededEventHandler = (
+	request: Express.Request,
+	response: Express.Response,
+	next: Express.NextFunction,
+	optionsUsed: Options,
+) => void
+
+/**
+ * Event callback that is triggered on a client's first request that exceeds the limit
+ * but not for subsequent requests. May be used for logging, etc. Should *not*
+ * send a response.
+ *
+ * @param request {Express.Request} - The Express request object
+ * @param response {Express.Response} - The Express response object
+ * @param optionsUsed {Options} - The options used to set up the middleware
+ */
+export type RateLimitReachedEventHandler = (
+	request: Express.Request,
+	response: Express.Response,
+	optionsUsed: Options,
+) => void
+
+/**
+ * Data returned from the `Store` when a client's hit counter is incremented.
+ *
+ * @property totalHits {number} - The number of hits for that client so far
+ * @property resetTime {Date | undefined} - The time when the counter resets
+ */
+export type IncrementResponse = {
+	totalHits: number
+	resetTime: Date | undefined
+}
+
+/**
+ * A modified Express request handler with the rate limit functions.
+ */
+export type RateLimitRequestHandler = Express.RequestHandler & {
+	/**
+	 * Method to reset a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	resetKey: (key: string) => void
+}
+
+/**
+ * An interface that all hit counter stores must implement.
+ *
+ * @deprecated 6.x - Implement the `Store` interface instead.
+ */
+export interface LegacyStore {
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 * @param callback {IncrementCallback} - The callback to call once the counter is incremented
+	 */
+	incr: (key: string, callback: IncrementCallback) => void
+
+	/**
+	 * Method to decrement a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	decrement: (key: string) => void
+
+	/**
+	 * Method to reset a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	resetKey: (key: string) => void
+
+	/**
+	 * Method to reset everyone's hit counter.
+	 */
+	resetAll?: () => void
+}
+
+/**
+ * An interface that all hit counter stores must implement.
+ */
+export interface Store {
+	/**
+	 * Method that initializes the store, and has access to the options passed to
+	 * the middleware too.
+	 *
+	 * @param options {Options} - The options used to setup the middleware
+	 */
+	init?: (options: Options) => void
+
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 *
+	 * @returns {IncrementResponse} - The number of hits and reset time for that client
+	 */
+	increment: (key: string) => Promise<IncrementResponse> | IncrementResponse
```

Comment:
I'm wondering if we should just make this and the other APIs async instead of having two return types. The MemoryStore is the only one that will ever be synchronous. What do you think?
```suggestion
	async increment: (key: string) => IncrementResponse
```

Reply:
The current type declaration (`Promise<T> | T`) allows the function to be both synchronous and asynchronous. We could remove the `| T` part to make it only asynchronous.

Also, we cannot add the `async` keyword to types - instead, an async function is declared by wrapping the return type with `Promise<...>` - https://stackoverflow.com/a/55890828.

Reply:
Aah, ok, I hadn't realized that. Let's just make it always return a promise then.

Reply:
Ok, will do

Reply:
Done!

Reply:
I just realised that if we force all the methods (especially `decrement`, `resetKey` and `resetAll`) to return a `Promise`, then other stores won't be able to maintain backwards compatibility - all the current functions are sync and return just `void`.

Reply:
aah, shucks. You're right.

There might be a way to make it work, 5.x doesn't care about the return type, let me think about it a bit...